### PR TITLE
Unified fetching/syncing block metadata everywhere 

### PR DESCRIPTION
**Pull request by lx223**
* [] CHANGELOG entry if change is relevant to the end user.

## Changes

* add a new block meta fetcher that fetches block metadata stored in a bucket
* replace some of the existing block md fetching code with the new fetcher
* change the signature of `DownloadMeta` to use the more restrictive `BucketReader`

## Verification

Unit test

### CODE REVIEW

**Comment by lx223** [comment]
The CI test fails because of go mod error. It doesn't seem related to my change. Is it a known issue?

**Comment by bwplotka** [comment]
Fixes https://github.com/thanos-io/thanos/issues/1335

@lx223 can you rebase, we fixed this master go.mod error (:

File: pkg/block/mdfetcher.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		concurrencyLimit: concurrencyLimit,
+		rwLock:           &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.
+func (f *MetadataFetcher) Fetch(ctx context.Context) (map[ulid.ULID]*metadata.Meta, error) {
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: Why fetcher and not e.g `Syncer`? Especially if we maintain `cachedBlockMDs` for efficiency?

Comment (bwplotka) [reply pkg/block/mdfetcher.go]: Let's mention if it's goroutine safe  (: It looks like it is, the question is ... should it be?

I think it would be better if you would not make `Fetch/Sync` goroutine safe. First of all, I don't think it is needed. Secondly, with this, we can be more efficient in terms of memory. We can allow to reuse exactly the same `map[ulid.ULID]*metadata.Meta` (pass as an argument and return updated one). This might save some allocations as there can be millions of objects potentially in the bucket. 

What do you think? With this I am not sure if the struct is needed, maybe just Function is enough? Don't have a strong opinion.

File: pkg/block/mdfetcher.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	rwLock         *sync.RWMutex
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: Usually we just name it `mu`. `rw` is part of type so no need to put type info into variable. (: 

```suggestion
	mu         *sync.RWMutex
```

File: pkg/block/mdfetcher.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
+package block
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: `syncer`?

Same for file `mdsyncer`?

File: pkg/block/mdfetcher.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	rwLock         *sync.RWMutex
+
+	concurrencyLimit int
+	logger           log.Logger
+	bkt              objstore.Bucket
+}
+
+func NewMetadataFetcher(lg log.Logger, concurrencyLimit int, bkt objstore.Bucket) *MetadataFetcher {
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]:
```suggestion
func NewMetadataFetcher(logger log.Logger, concurrencyLimit int, bkt objstore.Bucket) *MetadataFetcher {
```
Super minor nit, but `logger` is consistent across codebase.

File: pkg/block/mdfetcher.go (hunk 1 above)
Comment (bwplotka) [reply pkg/block/mdfetcher.go]: Be aware of this https://github.com/thanos-io/thanos/pull/1408 that changes this code as well (somehow we need to sync both PRs) cc @povilasv

File: pkg/block/mdfetcher.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		logger:           lg,
+		concurrencyLimit: concurrencyLimit,
+		rwLock:           &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: > If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.

1. `s/metadta/metadata`
2. Nice idea to delegate the no-meta-json blocks on caller side. I think it makes sense.

However, maybe two maps make sense here then? 
```
(metas map[ulid.ULID]*metadata.Meta, noMeta []ulid.ULID, err  error)
```
This will be easier to use on caller side I think.

File: pkg/block/mdfetcher.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+func (f *MetadataFetcher) Fetch(ctx context.Context) (map[ulid.ULID]*metadata.Meta, error) {
+	blockIDs := make(chan ulid.ULID, f.concurrencyLimit)
+	defer close(blockIDs)
+	remoteBlockMDs := make(map[ulid.ULID]*metadata.Meta, len(f.cachedBlockMDs))
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < f.concurrencyLimit; i++ {
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: Will review body of this function when we resolve other comments first (: 

Looks good from first glance, looks like you covered all the edge cases, nice! Thanks for tests as well!

**Comment by bwplotka** [comment]
Also we need CHANGELOG for this I think as we not handle partial uploads carefully on downsampling/retention apply

**Comment by bwplotka** [comment]
Hi :wave: 

Can we finish this? (: If you don't have time, can I grab it and finish?

File: pkg/block/mdfetcher.go (hunk 1 above)
Comment (lx223) [reply pkg/block/mdfetcher.go]: I am happy to rename to syncer if you feel strongly about it. I am alright with both. I chose fetch as I felt it is more about fetching the metadata from remote source. Keeping a cache is just an internal optimisation and might become an issue and get removed if the bucket contains too many objects.

Good point about about the memory allocation. I have a different perspective. I don't think it would add that much to memory as it is using metadata pointers right now. And if we were to reuse the same map, there would still have to be a map to keep what metadata are in the bucket in order to remove those cached but somehow-removed-from-remote-source ones. So the saving won't be that much. Besides, I think streaming metadata might be a better solution to scale with bigger number of objects in the bucket. What do you think?

File: pkg/block/mdfetcher.go (hunk 5 above)
Comment (lx223) [reply pkg/block/mdfetcher.go]: sorry, not following about point 1. what is `s/metadta/metadata`?

I thought about whether to distinguish those two cases. For the places I have replaced, I haven't found a case that would require these two being distinguished. Did I miss a case that would require this?

**Comment by lx223** [comment]
@bwplotka  sorry about that! 

I'll change `pkg/store/bucket.go` and add a changelog tomorrow.

File: pkg/block/mdfetcher.go (hunk 5 above)
Comment (bwplotka) [reply pkg/block/mdfetcher.go]: > sorry, not following about point 1. what is s/metadta/metadata?

https://www.tutorialspoint.com/unix/unix-regular-expressions.htm

(:

**Comment by bwplotka** [comment]
@lx223 Are you on CNCF slack? Join us so we can discuss quicker offline!

I would like to see this change ASAP as it will help with: https://github.com/thanos-io/thanos/issues/1331

File: pkg/block/mdfetcher.go (hunk 5 above)
Comment (bwplotka) [reply pkg/block/mdfetcher.go]: Hm, what about the case where we remove partially uploaded blocks? (noMeta + delay):

https://github.com/thanos-io/thanos/blob/865d5ec710b5651ead08a7dc73864a6f5dbeeaa9/pkg/compact/compact.go#L197

I think that's when noMeta is useful.

File: pkg/block/mdfetcher.go (hunk 1 above)
Comment (bwplotka) [reply pkg/block/mdfetcher.go]: No strong feeling to the naming. It really depends on what exactly interface we are going to apply as you are exploring streaming suggestion now (: So then `ForeachMeta` might make sense more?

>  I think streaming metadata might be a better solution to scale with bigger number of objects in the bucket. What do you think?

Definitely, however, this means we would need to revert to any of Iter concepts like

* passing a function to invoke for each meta
* `Iter` interface: 
```
// Iterator is a simple iterator that can only get the next value.
type Iterator interface {
	At() metadata.Meta)
	Err() error
	Next() bool
}
```

**Comment by bwplotka** [comment]
Would love to see this work to be done as this would enormously help with dynamic system where components are restarted in different moments causing partial uploads to be very common (: Let me know if I can help with this and I would love to prioritize this.  @lx223 

I listed all the places were we can unify our sync/fetch logic.

**Comment by bwplotka** [comment]
Relaed https://github.com/thanos-io/thanos/pull/1525

File: pkg/block/mdfetcher.go (hunk 5 above)
Comment (lx223) [reply pkg/block/mdfetcher.go]: I might be wrong here but I think this is the only case that this required and it is covered in the caller's logic - `thanos/pkg/compact/compact.go`. But it does sound like something useful in future. Tempted to add when we are seeing more such usages.

File: pkg/block/mdfetcher.go (hunk 3 above)
Comment (lx223) [reply pkg/block/mdfetcher.go]: resolved as it is related to the name discussion.

**Comment by lx223** [comment]
@bwplotka just joined CNCF slack and have made changes to all those places you listed. PTAL

File: pkg/block/mdfetcher.go (hunk 1 above)
Comment (lx223) [reply pkg/block/mdfetcher.go]: I think we don't have to use stream for now unless the OOM issue is already big and urgent. I think we can keep this signature which meshes pretty well with the current usage and we can use this as the foundation to identify and make changes when we are ready to explore other options like streaming.

File: cmd/thanos/bucket.go [hunk 7]
```diff
@@ -205,39 +207,41 @@ func registerBucketLs(m map[string]setupFunc, root *kingpin.CmdClause, name stri
...
 				fmt.Fprintln(os.Stdout, "")
 				return nil
 			}
 		}
 
-		return bkt.Iter(ctx, "", func(name string) error {
-			id, ok := block.IsBlockDir(name)
-			if !ok {
-				return nil
+		mdFetcher := block.NewMetadataFetcher(logger, defaultBlockMDFetchConcurrency, bkt)
```
Comment (lx223) [inline cmd/thanos/bucket.go]: didn't switch to the `download` function because the original logic suggests it requires block id even for partially uploaded blocks.

**Comment by bwplotka** [comment]
Should we get back to this @lx223 ? (:

**Comment by lx223** [comment]
@bwplotka sure :D WDYT about the current code? there is some rebasing to do. But, it probably won't change the main logic.

File: cmd/thanos/bucket.go (hunk 7 above)
Comment (bwplotka) [reply cmd/thanos/bucket.go]: :+1:

File: pkg/block/block.go [hunk 8]
```diff
@@ -163,7 +163,7 @@ func deleteDir(ctx context.Context, logger log.Logger, bkt objstore.Bucket, dir
 
 // DownloadMeta downloads only meta file from bucket by block ID.
 // TODO(bwplotka): Differentiate between network error & partial upload.
-func DownloadMeta(ctx context.Context, logger log.Logger, bkt objstore.Bucket, id ulid.ULID) (metadata.Meta, error) {
+func DownloadMeta(ctx context.Context, logger log.Logger, bkt objstore.BucketReader, id ulid.ULID) (metadata.Meta, error) {
```
Comment (bwplotka) [inline pkg/block/block.go]: :+1:

File: pkg/block/mdfetcher.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+	f.mu.RLock()
+	existingMD, exist := f.cachedBlockMDs[blockID]
+	f.mu.RUnlock()
+
+	if exist {
+		return existingMD, nil
+	}
+
+	md, err := DownloadMeta(ctx, f.logger, f.bkt, blockID)
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: Wonder if we can leverage the file system as a cache, instead of downloading and keep in memory? Let's maybe add TODO at least (: It saves some bandwidth (?) on startup.

File: pkg/block/mdfetcher.go (hunk 1 above)
Comment (bwplotka) [reply pkg/block/mdfetcher.go]: Yea, I think I am happy with this. We already keep `cached block MD` anyway.

Comment (bwplotka) [reply pkg/block/mdfetcher.go]: fetcher might be fine then

File: cmd/thanos/bucket.go [hunk 10]
```diff
@@ -383,25 +373,21 @@ func refresh(ctx context.Context, logger log.Logger, bucketUI *ui.Bucket, durati
 	})
 }
 
-func download(ctx context.Context, logger log.Logger, bkt objstore.Bucket) (blocks []metadata.Meta, err error) {
+func download(ctx context.Context, logger log.Logger, bkt objstore.Bucket) (blocks []*metadata.Meta, err error) {
 	level.Info(logger).Log("msg", "synchronizing block metadata")
+	mdFetcher := block.NewMetadataFetcher(logger, defaultBlockMDFetchConcurrency, bkt)
+	mds, err := mdFetcher.Fetch(ctx)
+	if err != nil {
+		level.Error(logger).Log("err", err, "msg", "Failed to downloaded block metadata")
```
Comment (bwplotka) [inline cmd/thanos/bucket.go]: double error handling, please choose one

File: pkg/block/mdfetcher.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	defer close(blockIDs)
+	remoteBlockMDs := make(map[ulid.ULID]*metadata.Meta, len(f.cachedBlockMDs))
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < f.concurrencyLimit; i++ {
+		go func() {
+			for blockID := range blockIDs {
+				md, err := f.fetchBlockMDIfNotCached(ctx, wg, blockID)
+				if err != nil {
+					level.Debug(f.logger).Log("msg", "failed to fetch block metadata", "block", blockID, "error", err)
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: warning?

File: pkg/store/bucket.go [hunk 12]
```diff
@@ -295,59 +294,38 @@ func (s *BucketStore) Close() (err error) {
...
 		}
-		return nil
-	})
-
-	close(blockc)
-	wg.Wait()
 
-	if err != nil {
-		return errors.Wrap(err, "iter")
+		if err := s.addBlock(ctx, blockID); err != nil {
```
Comment (bwplotka) [inline pkg/store/bucket.go]: This is important as we download meta.json twice. Once to memory once to disk... Can we improve this?

File: pkg/block/mdfetcher.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+		case blockIDs <- id:
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	wg.Wait()
+
+	f.mu.Lock()
```
Comment (bwplotka) [inline pkg/block/mdfetcher.go]: Why this lock? To make fetcher thread-safe? I think we need single flight here TBH with some expiry even (: Can you add `TODO(bwplotka): Add single flight with expiry if needed`

Also, I am not sure if we can really reuse this lock IMO...  I would rather have:
either 
* two locks, one for concurrency here and second for potential concurrent use of `Fetch` 
* remove this lock. I don't think we do currently fetch concurrently so we could remove this and comment that it's not thread safe

I would say let's add `TODO`

File: pkg/block/mdfetcher.go (hunk 11 above)
Comment (squat) [reply pkg/block/mdfetcher.go]: I think we should collect and/or exit on errors. Currently, if there was a transient network error we'll cache `nil` (negative result) and then never re-try downloading that metadata.

File: pkg/block/mdfetcher.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	for i := 0; i < f.concurrencyLimit; i++ {
+		go func() {
+			for blockID := range blockIDs {
+				md, err := f.fetchBlockMDIfNotCached(ctx, wg, blockID)
+				if err != nil {
+					level.Debug(f.logger).Log("msg", "failed to fetch block metadata", "block", blockID, "error", err)
+				}
+
+				f.mu.Lock()
+				remoteBlockMDs[blockID] = md
```
Comment (squat) [inline pkg/block/mdfetcher.go]: I think there is a race currently: we call `wg.Done` before inserting the returned values in the map, meaning `wg.Done()` may return early and we may return an incomplete map to the caller.

**Comment by bwplotka** [comment]
@lx223 let us know if you have time to finish this. This fixes quite annoying issue, plus @squat already touching this path for compactor syncer. 

If not we are happy to continue this for you (:

**Comment by lx223** [comment]
@bwplotka  sry, might have enough time to finish it fast. Feel free to take it. :)

**Comment by bwplotka** [comment]
Cool, closing this for now then, updating the issue.
